### PR TITLE
Add EM Boundaries Generalist

### PR DESCRIPTION
**Pull request by anwai98**

### CODE REVIEW

File: finetuning/generalists/training/electron_microscopy/boundaries/obtain_boundaries_em_datasets.py [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+        rois[sample_id] = roi
+    return rois
+
+
+def axondeepseg_label_trafo(labels):
+    # after checking, labels look like this : 0 is bg, 1 is myelins and 2 is axons
+    foreground_seeds = label((labels == 2))
+    boundary_prediction = (labels == 1)
+    seg = watershed(boundary_prediction, markers=foreground_seeds, mask=(foreground_seeds + boundary_prediction) > 0)
+    seg = label_consecutive(seg)
```
Comment (constantinpape) [inline finetuning/generalists/training/electron_microscopy/boundaries/obtain_boundaries_em_datasets.py:37]: Applying label consecutive here should not be necessary. The seeds and corresponding watershed will already be consecutive.

File: finetuning/generalists/training/electron_microscopy/boundaries/obtain_boundaries_em_datasets.py [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+        ) for coord in valid_coordinates)
+        rois[sample_id] = roi
+    return rois
+
+
+def axondeepseg_label_trafo(labels):
+    # after checking, labels look like this : 0 is bg, 1 is myelins and 2 is axons
+    foreground_seeds = label((labels == 2))
+    boundary_prediction = (labels == 1)
+    seg = watershed(boundary_prediction, markers=foreground_seeds, mask=(foreground_seeds + boundary_prediction) > 0)
```
Comment (constantinpape) [inline finetuning/generalists/training/electron_microscopy/boundaries/obtain_boundaries_em_datasets.py:36]: Did you (visually!) check the results for this? Running the watershed with a binary height map could give weird results for touching myelinated axons. It could be better to run this on some distance map (we can discuss this tomorrow).

File: finetuning/generalists/training/electron_microscopy/boundaries/obtain_boundaries_em_datasets.py (hunk 1 above)
Comment (anwai98) [reply finetuning/generalists/training/electron_microscopy/boundaries/obtain_boundaries_em_datasets.py:37]: True, makes sense. Thanks!

File: finetuning/generalists/training/electron_microscopy/boundaries/obtain_boundaries_em_datasets.py (hunk 2 above)
Comment (anwai98) [reply finetuning/generalists/training/electron_microscopy/boundaries/obtain_boundaries_em_datasets.py:36]: Oh yeah, true. It's hard to say for these touching ones. For most, does fine, for some, the touching objects do look weird.

Comment (constantinpape) [reply finetuning/generalists/training/electron_microscopy/boundaries/obtain_boundaries_em_datasets.py:36]: Yeah, that's what I would expect. The best approach for this is usually to use a distance transform as the height map, then it will assign the pixels to the closest seed, which should probably give the best result.
Here's the rough idea (!not tested):
```python
from scipy.ndimage import distance_transform_edt

# use the distance to the axons as height map to assign pixels to nearest axon
hmap = distance_transform_edt(labels != 1)
seg = watershed(hmap, ...)
```

**Comment by anwai98** [comment]
Looks like it's fixed now. I updated the calculation of the distances to the whole myelinated axons, then it seems to work. (there is some minor flooding leakage in larger neighbouring objects, not sure if it's critical)

File: finetuning/generalists/training/electron_microscopy/boundaries/obtain_boundaries_em_datasets.py [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+
+
+def axondeepseg_label_trafo(labels):
+    # after checking, labels look like this : 0 is bg, 1 is myelins and 2 is axons
+    foreground_seeds = label((labels == 2))
+    boundary_prediction = (labels == 1)
+
+    # use the distance to the myelinated axons as height map to assign pixels to nearest (myelinated axon)
+    hmap = distance_transform_edt(labels == 0)
+    seg = watershed(image=hmap, markers=foreground_seeds, mask=(foreground_seeds + boundary_prediction) > 0)
```
Comment (anwai98) [inline finetuning/generalists/training/electron_microscopy/boundaries/obtain_boundaries_em_datasets.py:39]: @constantinpape Here's the updated distances-based watershed transform. I left the napari scripts uncommented, in case you want to double check the watershed outputs.

File: finetuning/generalists/training/electron_microscopy/boundaries/obtain_boundaries_em_datasets.py [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+    return rois
+
+
+def axondeepseg_label_trafo(labels):
+    # after checking, labels look like this : 0 is bg, 1 is myelins and 2 is axons
+    foreground_seeds = label((labels == 2))
+    boundary_prediction = (labels == 1)
+
+    # use the distance to the myelinated axons as height map to assign pixels to nearest myelinated axon
+    hmap = distance_transform_edt(labels == 0)
```
Comment (constantinpape) [inline finetuning/generalists/training/electron_microscopy/boundaries/obtain_boundaries_em_datasets.py:38]:
```suggestion
    hmap = distance_transform_edt(labels != 2)
```

Comment (anwai98) [reply finetuning/generalists/training/electron_microscopy/boundaries/obtain_boundaries_em_datasets.py]: Oh yeah. okay okay. True. Let's stick to this then. Thanks!

File: finetuning/generalists/training/electron_microscopy/boundaries/obtain_boundaries_em_datasets.py [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+    # creating weights for each concatenated dataset
+    train_weights = np.concatenate(
+        tuple([np.linspace(1, 3, len(_dataset)) for _dataset in train_datasets])
+    )
+    val_weights = np.concatenate(
+        tuple([np.linspace(1, 3, len(_dataset)) for _dataset in val_datasets])
+    )
+
+    train_sampler = WeightedRandomSampler(train_weights, len(generalist_em_train_dataset))
+    val_sampler = WeightedRandomSampler(val_weights, len(generalist_em_val_dataset))
```
Comment (anwai98) [inline finetuning/generalists/training/electron_microscopy/boundaries/obtain_boundaries_em_datasets.py:129]: @constantinpape I scratched up an initlal setup for upsampling the AxonDeepSeg dataset. Works decent. (maybe you have another suggestion on this?)

Comment (constantinpape) [reply finetuning/generalists/training/electron_microscopy/boundaries/obtain_boundaries_em_datasets.py:129]: You could also just set `n_samples` to something higher in the axondeepseg dataset.
You can pass `n_samples` as kwarg to it, which will then be taken by the segmentation dataset:
https://github.com/constantinpape/torch-em/blob/main/torch_em/data/segmentation_dataset.py#L34

Comment (constantinpape) [reply finetuning/generalists/training/electron_microscopy/boundaries/obtain_boundaries_em_datasets.py:129]: It just repeats them.

File: finetuning/generalists/training/electron_microscopy/boundaries/train_boundaries_em_generalist.py [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+            checkpoint_path=checkpoint_path,
+            model_type=model_type,
+            save_path=args.export_path,
+        )
+
+
+def main():
+    parser = argparse.ArgumentParser(description="Finetune Segment Anything for the Boundary Structures EM datasets.")
+    parser.add_argument(
+        "--input_path", "-i", default="/scratch/usr/nimanwai/data/",
```
Comment (constantinpape) [inline finetuning/generalists/training/electron_microscopy/boundaries/train_boundaries_em_generalist.py:101]: Can you update the path so that it's also pointing to the project folder?

**Review by constantinpape — COMMENTED** [review]
Looks good, but let's keep the data in one place.

File: finetuning/generalists/training/electron_microscopy/boundaries/train_boundaries_em_generalist.py (hunk 6 above)
Comment (anwai98) [reply finetuning/generalists/training/electron_microscopy/boundaries/train_boundaries_em_generalist.py]: Done :) (also updated the permissions). Thanks!
